File: gdscript/src/main/java/gdscript/competion/GdClassNameCompletionContributor.java
package gdscript.competion;

import com.intellij.codeInsight.completion.*;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiManager;
import com.intellij.psi.search.FileTypeIndex;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.util.PsiTreeUtil;
import gdscript.GdFileType;
import gdscript.GdIcon;
import gdscript.competion.staticLoader.StaticClassLoader;
import gdscript.competion.utils.CompletionPriority;
import gdscript.competion.utils.FileCompletionUtil;
import gdscript.competion.utils.PositionUtil;
import gdscript.psi.GdClassNaming;
import gdscript.psi.GdFile;
import gdscript.psi.GdTypes;
import gdscript.psi.impl.GdClassNamingImpl;
import gdscript.psi.utils.PsiGdFileUtil;
import org.jetbrains.annotations.NotNull;

public class GdClassNameCompletionContributor extends CompletionContributor {

    @Override
    public void fillCompletionVariants(@NotNull CompletionParameters parameters, @NotNull CompletionResultSet resultSet) {

        var project = parameters.getPosition().getProject();
        var  virtualFiles =
                FileTypeIndex.getFiles(GdFileType.INSTANCE, GlobalSearchScope.allScope(project));
        for (VirtualFile it : virtualFiles) {
                var gdf = (GdFile) PsiManager.getInstance(project).findFile(it);
                var properties = (GdClassNamingImpl) PsiTreeUtil.findChildOfType(gdf, GdClassNamingImpl.class);
                var ss = properties.getGreenStub();
                var asd=  546;
            }

        GdClassNaming ar = (GdClassNaming) parameters.getPosition().getParent().getParent();

        var asd = ar.getClassname();

        // After EXTENDS keyword
        if (PositionUtil.isLeafPreceded(parameters, GdTypes.EXTENDS)) {
            String myName = PsiGdFileUtil.INSTANCE.myFilename(parameters);
            PsiGdFileUtil.INSTANCE.customs(parameters)
                    .forEach(name -> {
                                if (!name.equals(myName)) {
                                    resultSet.addElement(
                                            PrioritizedLookupElement.withPriority(
                                                    LookupElementBuilder
                                                            .create(name)
                                                            .withIcon(GdIcon.OBJECT),
                                                    CompletionPriority.USER_DEFINED
                                            ));
                                }
                            }
                    );
            StaticClassLoader
                    .getClasses()
                    .forEach((_key, gd) -> resultSet.addElement(
                            PrioritizedLookupElement.withPriority(
                                    LookupElementBuilder
                                            .create(gd.name)
                                            .withIcon(GdIcon.getEditorIcon(gd.name)),
                                    CompletionPriority.STATIC_TOP
                            )));

        // After CLASS_NAME keyword
        } else if (PositionUtil.isLeafPreceded(parameters, GdTypes.CLASS_NAME)) {
            GdFile file = FileCompletionUtil.currentFile(parameters);
            String name = file.getName(); // remove .gd suffix
            resultSet.addElement(
                    LookupElementBuilder.create(name.substring(0, name.length() - 3))
            );
        }

        super.fillCompletionVariants(parameters, resultSet);
    }

}

File: gdscript/src/main/kotlin/gdscript/completion/GdClassNameCompletionContributor.kt
package gdscript.completion

import com.intellij.codeInsight.completion.CompletionContributor
import com.intellij.codeInsight.completion.CompletionParameters
import com.intellij.codeInsight.completion.CompletionResultSet
import com.intellij.patterns.PlatformPatterns.psiElement
import gdscript.GdKeywords
import gdscript.psi.GdClassNaming
import gdscript.utils.StringUtil.snakeToPascalCase

/**
 * Filename to class_name completion
 */
class GdClassNameCompletionContributor : CompletionContributor() {

    val CLASS_NAME_NM = psiElement()
        .afterLeaf(psiElement().withText(GdKeywords.CLASS_NAME))
        .withSuperParent(2, GdClassNaming::class.java)

    override fun fillCompletionVariants(parameters: CompletionParameters, result: CompletionResultSet) {
        if (CLASS_NAME_NM.accepts(parameters.position)) {
            val filename = parameters.originalFile.name
            result.addElement(
                GdLookup.create(filename.substring(0, filename.lastIndexOf(".")).snakeToPascalCase()),
            )
            result.stopHere()
        }
    }

}
